client/console/src/main/java/org/apache/syncope/client/console/audit/AuditHistoryDirectoryPanel.java

```diff
@@ -0,0 +1,316 @@
+        switch (this.anyTypeKind) {
+            case USER:
+                auditRestoreEntitlement = StandardEntitlement.USER_UPDATE;
+                break;
+            case GROUP:
+                auditRestoreEntitlement = StandardEntitlement.GROUP_UPDATE;
+                break;
+            default:
+                auditRestoreEntitlement = StandardEntitlement.ANYTYPE_UPDATE;
+                break;
+        }
+
+        panel.add(new ActionLink<AnyTOAuditEntryBean>() {
+
+            private static final long serialVersionUID = -6745431735457245600L;
+
+            @Override
+            public void onClick(final AjaxRequestTarget target, final AnyTOAuditEntryBean modelObject) {
+                try {
+                    AuditHistoryDirectoryPanel.this.getTogglePanel().close(target);
+                    restore(modelObject, anyTypeKind, anyTO);
+                    target.add(container);
+                } catch (SyncopeClientException e) {
+                    LOG.error("While restoring {}", auditEntryTO.getEntityKey(), e);
+                    SyncopeConsoleSession.get().error(StringUtils.isBlank(e.getMessage())
+                        ? e.getClass().getName() : e.getMessage());
+                }
+                ((BasePage) pageRef.getPage()).getNotificationPanel().refresh(target);
+            }
+        }, ActionLink.ActionType.RESTORE, auditRestoreEntitlement);
+
+        return panel;
+    }
+
+    @Override
+    protected Collection<ActionLink.ActionType> getBatches() {
+        return Collections.emptyList();
+    }
+
+    /**
+     * Restore an object based on the audit record.
+     * <p>
+     * Note that for user objects, the original audit record masks
+     * the password and the security answer; so we cannot use the audit
+     * record to resurrect the entry based on mask data. The method behavior
+     * below will reset the audit record such that the current security answer
+     * and the password for the object are always maintained, and such properties
+     * for the user cannot be restored using audit records.
+     *
+     * @param entryBean   the entry bean
+     * @param anyTypeKind the any type kind
+     * @param anyTO       the any to
+     * @return the response
+     */
+    private ProvisioningResult<? extends AnyTO> restore(final AnyTOAuditEntryBean entryBean,
+                                                        final AnyTypeKind anyTypeKind, final AnyTO anyTO) {
+        try {
+            if (anyTypeKind == AnyTypeKind.USER) {
+                UserTO userTO = (UserTO) MAPPER.readValue(entryBean.getBefore().toString(), anyTypeKind.getTOClass());
+                UserTO current = new UserRestClient().read(anyTO.getKey());
+                userTO.setPassword(((UserTO) anyTO).getPassword());
+                userTO.setSecurityAnswer(((UserTO) anyTO).getSecurityAnswer());
+                UserPatch userPatch = AnyOperations.diff(userTO, current, false);
```

Comment:
After obtaining `userPatch`, add the following statements:

```java
userPatch.setPassword(null);
userPatch.setSecurityAnswer(null);
```

to ensure that no security-sensible changes are actually performed during restore.